shardy/dialect/sdy/ir/attrs.td

```diff
@@ -746,6 +790,17 @@ def Sdy_OpShardingRule : AttrDef<Sdy_Dialect, "OpShardingRule"> {
     these ops, so a user must tell it how. When it is a custom rule, then the
     rule is always preserved/never removed. `is_custom_rule` can only be true
     for `stablehlo.custom_call` ops.
+
+    **Constraints:**
+    - Elements in `reduction_factors` and `need_replication_factors` must
```

Comment:
This sounds a bit off, the only constraint of `DimMappingAttr` that is relevant is:
"Factor indices must be in range [0, `$factor_sizes`)" so lets put it explicitly

Another constraint that is specific to `reduction_factors` and `need_replication_factors`  is: there shouldn't be duplicate factor indices in each one, and also across the two.

Reply:
Lets put constraints for those at the bottom

Reply:
Maybe my sentence is redundant now? I just wanted to make it clear. Feel free to suggest other wordings

Reply:
It's not redundant, you are right that we should add constraints for `reduction_factors` and `need_replication_factors`, they should just be detached from DimMappingAttr.

See my suggestions above.